index.js

```diff
@@ -1,77 +1,184 @@
+    this.unwatch = null
 
-  fs.lstat(name, function (_, st) {
-    if (!st || stopped) {
-      stopped = true
-      return
+    this.opened = false
+    this.closed = false
+
+    this._opening = this._ready()
+    this._closing = null
+
+    this._opening.catch(safetyCatch)
+  }
+
+  ready () {
+    return this._opening
+  }
+
+  async _ready () {
+    const st = await fs.promises.lstat(this.filename)
+
+    if (this.closed) return
+
+    const watch = st.isDirectory() ? watchDirectory : watchFile
+    this.unwatch = watch(this.filename, this.onchange)
+
+    this.opened = true
+  }
+
+  async close () {
+    if (this._closing) return this._closing
+    this._closing = this._close()
+    return this._closing
+  }
+
+  async _close () {
+    if (this.closed) return
+    this.closed = true
+
+    if (!this.opened) await this._opening.catch(safetyCatch)
+
+    if (this.unwatch) {
+      await this.unwatch()
+      this.unwatch = null
     }
-    clear = st.isDirectory() ? watchDirectory(name, onchange) : watchFile(name, onchange)
-  })
+  }
+}
+
+class Watch {
+  constructor (filename, opts, onchange) {
+    this.filename = filename
+    this.opts = opts || {}
+    this.onchange = onchange
+
+    this.opened = true
+    this.closed = false
 
-  return function () {
-    if (stopped) return
-    stopped = true
-    if (clear) {
-      clear()
+    this._closing = null
+
+    this.watcher = fs.watch(this.filename, this.opts, this.onchange)
+    this.watcherClosed = false
+
+    this.watcher.on('error', noop)
+    this.watcher.on('close', () => {
+      this.watcherClosed = true
+    })
+  }
+
+  async close () {
+    if (this._closing) return this._closing
+    this._closing = this._close()
+    return this._closing
+  }
+
+  async _close () {
+    if (this.closed) return
+    this.closed = true
+
+    let onclose = null
+
+    if (!this.watcherClosed) {
+      onclose = new Promise(resolve => this.watcher.on('close', resolve))
     }
+
+    this.watcher.close()
+    await onclose
+
+    this.watcher = null
   }
 }
 
 function watchFile (filename, onchange) {
-  var prev = null
-  var prevTime = 0
+  let prev = null
+  let prevTime = 0
+  let actives = 0
+
+  let closed = false
+  let close = null
+  const closing = new Promise(resolve => { close = resolve })
+
+  const watcher = new Watch(filename, null, function () {
+    actives++
 
-  var w = fs.watch(filename, function () {
     fs.lstat(filename, function (_, st) {
-      var now = Date.now()
+      actives--
+
+      if (closed) {
+        oncleanup()
+        return
+      }
+
+      const now = Date.now()
       if (now - prevTime > 2000 || !same(st, prev)) onchange(filename)
       prevTime = now
       prev = st
     })
   })
 
-  return function () {
-    w.close()
+  return async function unwatch () {
+    if (closed) return closing
+    closed = true
+
+    await watcher.close()
+    oncleanup()
+
+    return closing
+  }
+
+  function oncleanup () {
+    if (actives === 0) close()
```

Comment:
Sanity check: is it guaranteed that, once actives is decreased to 0, it will never increment again? (as in: once `unwatch` is called, the `onchange` will never be called again)

Reply:
Yep guaranteed. The first part of the `lstat` callback might get executed but it won't continue due `closed` being true as soon as you called `unwatch`, so no extras `onchange`

Reply:
This was part of the cleanup leaks